File: raspberry-pi-ios-client/RaspberryPiClient/Data/Repository/Turtlesim/Spawn/TurtlesimServiceArgs.swift
//
//  TurtlesimServiceArgs.swift
//  RaspberryPiClient
//
//  Created by Shusuke Ota on 2025/12/28.
//

import Foundation

struct TurtlesimServiceArgs: RosCallServiceArgsProtocol, RosCallServiceCdrEncodable {
    let x: Float
    let y: Float
    let theta: Float
    let name: String?

    func encodeCdr() -> Data {
        var data = Data(CdrHelpers.encapsulationHeader)
        data.append(contentsOf: withUnsafeBytes(of: x.bitPattern.littleEndian) { Array($0) })
        data.append(contentsOf: withUnsafeBytes(of: y.bitPattern.littleEndian) { Array($0) })
        data.append(contentsOf: withUnsafeBytes(of: theta.bitPattern.littleEndian) { Array($0) })
        data.append(contentsOf: CdrHelpers.encodeCdrString(name))
        return data
    }
}

File: raspberry-pi-android-client/app/src/main/java/com/shusuke/raspberry_pi_client/data/repository/turtlesim/TurtlesimServiceArgs.kt
package com.shusuke.raspberry_pi_client.data.repository.turtlesim

import com.shusuke.raspberry_pi_client.data.infrastructure.ros.cdr.CdrHelpers
import kotlinx.serialization.Serializable
import java.nio.ByteBuffer
import java.nio.ByteOrder

/**
 * turtlesim の `/spawn` サービスの引数
 */
@Serializable
data class TurtlesimServiceArgs(
    val x: Float = 0f,
    val y: Float = 0f,
    val theta: Float = 0f,
    val name: String? = null
) {
    companion object {
        /**
         * CDR LE バイナリにエンコードする。
         *
         * 構造: header(4) + x(4) + y(4) + theta(4) + name(CDR string)
         */
        fun encodeToCdr(args: TurtlesimServiceArgs): ByteArray {
            val nameEncoded = CdrHelpers.encodeCdrString(args.name ?: "")
            val buffer = ByteBuffer.allocate(4 + 4 + 4 + 4 + nameEncoded.size)
                .order(ByteOrder.LITTLE_ENDIAN)
            buffer.put(CdrHelpers.encapsulationHeader)
            buffer.putFloat(args.x)
            buffer.putFloat(args.y)
            buffer.putFloat(args.theta)
            buffer.put(nameEncoded)
            return buffer.array()
        }
    }
}
